quinn-udp/src/unix.rs

```diff
@@ -770,7 +907,16 @@ fn set_socket_option(
 
 const OPTION_ON: libc::c_int = 1;
 
-#[cfg(not(target_os = "linux"))]
+#[cfg(any(target_os = "macos", target_os = "ios"))]
+mod gro {
+    use super::BATCH_SIZE;
+
+    pub(super) fn gro_segments() -> usize {
+        BATCH_SIZE
```

Comment:
`quinn-udp` **does not** expose `sendmmsg` style sends, i.e. only supports multiple sends via GSO from a single buffer with equal size segments. Thus the above hack of returning `BATCH_SIZE` on `max_gso_segments`, despite the fact that MacOS does not support GSO, while not ideal, seems reasonable. The large multi-segment datagram received from the user is segmented in user-space and then passed to MacOS as individual `iovec`s, `sendmmsg` style.

`quinn-udp` **does** expose `recvmmsg` style reads, i.e. supports receiving multiple datagrams both in separate `iovec`s and via GRO. Given that MacOS does not support GRO, why return `BATCH_SIZE` here and not `1`?

```suggestion
        1
```

Reply:
So that `UdpSocketState::gro_segments` is initialized to something other than 1? What am I missing?

Reply:
> So that `UdpSocketState::gro_segments` is initialized to something other than 1?

Why does it need to be initialized to something other than `1`?

If `UdpSocketState::gro_segments` is `1`, a user can still provide multiple buffers to `UdpSocketState::recv` and, on the MacOS code-path, each of those buffers can be passed to the OS to be filled in a `recvmmsg` style.

Reply:
Put differently, the value of `UdpSocketState::gro_segments` only determines, how many segments a user can receive in one `iovec`, not how many `iovec`s a user can pass to `UdpSocketState::recv()`.

In the case of quinn as a user, it simply determines how large each `iovec` should be, i.e. how many segments each `iovec` should be able to fit:

https://github.com/quinn-rs/quinn/blob/9a656d2d0b3031390e71a24258b5432115cfa381/quinn/src/endpoint.rs#L745-L750

Note that `recv_buf` here represents all `iovec`s in sequence passed to `UdpSocketState::recv()`.

(Quinn maintainers, please correct me if I am missing something.)

Reply:
Yes. Thank you @larseggert.